### DISCUSSION TITLE
Is there a way to pan and set focus to a node by their id?

### DISCUSSION THREAD

**Discussion by sagunji** [xyflow/xyflow#3264] [Q&A] [answered]
I have a dynamic canvas which is updated with new nodes by user actions. I have positioned my nodes with auto-layout with dagre.
When a new node is added, I want the canvas to focus on that node i.e set it as center.
I tried using `fitView({ nodes: [...] })` and looking at the implementation, its enough to pass arrays of node ids.

```
const JourneysCanvas = ({
  nodes,
  edges,
  children,
  focusNodeId,
  onNodesChange,
  onNodesPositionChanged,
}: JourneysCanvasProps) => {
  const { getNodes, fitView } = useReactFlow();
  const dragged = useRef(false);
  
  useEffect(() => {
    if (focusNodeId) {
      fitView({ nodes: [{ id: focusNodeId }]});
    }
  }, [fitView, focusNodeId])
....
  return (
<ReactFlow
        proOptions={PRO_OPTIONS}
        onNodesChange={onNodesChange}
        nodes={nodes}
        edges={edges}
        nodeTypes={NODE_TYPES}
        edgeTypes={EDGE_TYPES}
        onNodeDrag={handleNodeDrag}
        onNodeDragStop={handleNodeDragStop}
        onlyRenderVisibleElements
        fitView
        multiSelectionKeyCode={null}
      >
....
  )
}
```

[b7012c09-a412-4106-b0ed-966fc7547230 (1).webm](https://github.com/wbkd/react-flow/assets/19725393/b1c3ed8d-ef49-42e4-9878-a8cd395bc62a)

But it doesn't seem to work.
Initially, I passed the whole node object as focusNode as props but useEffect is called when any property of the node changes.

Any suggestion or help?

**Comment by hayleigh-dot-dev** [comment]
What happens if you wrap the `fitView` call in an animation frame request like:
```js
  useEffect(() => {
    if (focusNodeId) {
      window.requestAnimationFrame(() => {
        fitView({ nodes: [{ id: focusNodeId }]});
      })
    }
  }, [fitView, focusNodeId])
```

**Reply by sagunji** [reply]
There is no impact. 
[ee4ba9a7-d7d5-4a58-bf2b-538895988ca5.webm](https://github.com/wbkd/react-flow/assets/19725393/b9d14bd3-8b55-4846-850f-aa9515eb4e26)

The code does run and once as expected. But it doesn't apply fitView.

```
const JourneysCanvas = ({
  nodes,
  edges,
  children,
  focusNodeId,
  onNodesChange,
  onNodesPositionChanged,
}: JourneysCanvasProps) => {
  const { getNodes, fitView } = useReactFlow();
  const dragged = useRef(false);

   useEffect(() => {
    if (focusNodeId) {
      window.requestAnimationFrame(() => {
        fitView({ nodes: [{ id: focusNodeId }]});
      })
    }
  }, [fitView, focusNodeId])

  const handleNodeDrag = useCallback(() => {
    dragged.current = true;
  }, []);

  const handleNodeDragStop = useCallback(() => {
    if (dragged.current) {
      const allNodes = getNodes().map(({ id, position }) => ({
        nodeId: id,
        x: position.x,
        y: position.y,
      }));

      onNodesPositionChanged?.(allNodes);
    }

    dragged.current = false;
  }, [getNodes, onNodesPositionChanged]);

  return (
    <>
      <CustomEdgeMarkerDefs />
      <ReactFlow
        proOptions={PRO_OPTIONS}
        onNodesChange={onNodesChange}
        nodes={nodes}
        edges={edges}
        nodeTypes={NODE_TYPES}
        edgeTypes={EDGE_TYPES}
        onNodeDrag={handleNodeDrag}
        onNodeDragStop={handleNodeDragStop}
        onlyRenderVisibleElements
        fitView
        multiSelectionKeyCode={null}
      >
        <Background />
        {children}
      </ReactFlow>
    </>
  );
};

export default JourneysCanvas;
```

This is the whole canvas configurations. The parent wrapper is as:
```
 <ReactFlowProvider>
        <JourneysCanvas
          nodes={nodes}
          edges={edges}
          focusNodeId={focusNodeId}
          onNodesChange={handleNodesChange}
          onNodesPositionChanged={onNodePositionChange}
        >
           {controls}
        </JourneysCanvas>
      </ReactFlowProvider>
```

Its pretty straight-forward and simple.

**Reply by bcakmakoglu** [reply]
You're calling fitView to early from what I can tell in your code.

I assume you are rendering the graph, passing a "focus node id" to it and trying to call `fitView`.

What is most likely happening is, you call `fitView` before the viewport helpers are actually ready and thus it will just run a `noop` function, doing nothing.

Either run your `fitView` call in `onInit={() => { ... }}` or check if the nodes have been initialised `const nodesInitialized = useNodesInitialized(options)` before calling `fitView`.

**Reply by bcakmakoglu** [reply]
[Here's a basic example using a button to focus specific nodes.](https://codesandbox.io/s/silent-snowflake-wt96mr?file=/src/Flow.tsx)

**Reply by sagunji** [reply]
I believe so, `fitView` is being called before the canvas is fully ready. Thank you for sharing the example, though the action is event driven where the instance is fully initialized. I have [updated the example ](https://codesandbox.io/s/nameless-haze-4jnrm5?file=/src/Flow.tsx) to show that if nodes are dynamically added, it doesn't seem to pan towards the node. Explicitly setting a timeout is a hack that I could think of.

I tried to work with `useNodesInitialized` but same result.
Similarly, about checking whether nodes are initialized, does `useNodesInitialized` work similar to 
```
const nodesInitializedSelector = (state: ReactFlowState) =>
   [...state.nodeInternals.values()].every((node: Node) => node.width && node.height);
```
If that is so, nodes are updated with a width and height prior to being passed on reactflow as I am apply auto-layout from parent component.

**Reply by bcakmakoglu** [reply, marked as answer]
I see.
There is a way of doing this, it might seem a bit un-intuitive and there's probably different ways of going about this (for example by checking if a specific node is initialized yet and then triggering fitView) *but* imo this is the "easiest" way of doing this.

In your  `onNodesChange` handler you can check for changes of type `dimensions`.
When a node is initialized this change will be fired with the actual dimensions of the node, which should be anything but 0,0.

So you can basically listen for that event and call fit view when it happens.

```ts
onNodesChange={(changes) => {
	// call the actual change handler to apply the node changes to your nodes
  onNodesChange(changes);

	// loop through the changes and check for a dimensions change that relates to the node we want to focus
  changes.forEach((change) => {
    if (
      change.type === "dimensions" &&
      focusId === change.id &&
      change.dimensions &&
      change.dimensions.height > 0 &&
      change.dimensions.width > 0
    ) {
      fitView({
        nodes: [{ id: focusId }],
        duration: 500
      });

		// reset the focus id so we don't retrigger fit view when the dimensions of this node happen to change
      setFocusId(null);
    }
  });
}}
```

[A working example can be found here.](https://codesandbox.io/s/great-river-xpxdz5?file=/src/Flow.tsx:2135-2167)

**Reply by sagunji** [reply]
This worked perfectly. I see the reasoning behind it. It seems to be a forceful but right now it behaves as expected. Thank you!

**Comment by amitguptagwl** [comment]
I've achieved this by creating a tag under `SvelteFlowProvider`. Now I can use `useSvelteFlow` and change view by calling `fitView` method. You can see demo at [Text2Chart](https://solothought.com/text2chart/flow). And code [here](https://github.com/solothought/text2chart/blob/main/src/lib/flow/ViewportAdjuster.svelte)
